src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
+         return false;
+      }
+      break;
+
+   case _mongoc_uri_finalize_allowed:
+   default:
+      break;
+   }
+
+   return true;
+}
+
+static bool
+_finalize_auth_mechanism_properties_prohibited (const bson_t *mechanism_properties,
+                                                const char *mechanism,
+                                                bson_error_t *error)
+{
+   BSON_OPTIONAL_PARAM (mechanism_properties);
+   BSON_ASSERT_PARAM (mechanism);
+   BSON_OPTIONAL_PARAM (error);
 
-   if (bson_iter_init_find_case (&iter, &uri->credentials, MONGOC_URI_AUTHSOURCE)) {
-      source = bson_iter_utf8 (&iter, NULL);
+   if (mechanism_properties) {
+      MONGOC_URI_ERROR (error, "'%s' authentication mechanism does not accept mechanism properties", mechanism);
+      return false;
    }
 
-   if (mongoc_uri_get_auth_mechanism (uri)) {
-      /* authSource with GSSAPI or X509 should always be external */
-      if (!strcasecmp (mongoc_uri_get_auth_mechanism (uri), "GSSAPI") ||
-          !strcasecmp (mongoc_uri_get_auth_mechanism (uri), "MONGODB-X509")) {
-         if (source) {
-            if (strcasecmp (source, "$external")) {
-               MONGOC_URI_ERROR (error, "%s", "GSSAPI and X509 require \"$external\" authSource");
+   return true;
+}
+
+typedef struct __supported_mechanism_properties {
+   const char *name;
+   bson_type_t type;
+} supported_mechanism_properties;
+
+static bool
+_supported_mechanism_properties_check (const supported_mechanism_properties *supported_properties,
+                                       const bson_t *mechanism_properties,
+                                       const char *mechanism,
+                                       bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (supported_properties);
+   BSON_ASSERT_PARAM (mechanism_properties);
+   BSON_ASSERT_PARAM (mechanism);
+   BSON_ASSERT_PARAM (error);
+
+   bson_iter_t iter;
+   BSON_ASSERT (bson_iter_init (&iter, mechanism_properties));
+
+   // For each element in `MongoCredential.mechanism_properties`...
+   while (bson_iter_next (&iter)) {
+      const char *const key = bson_iter_key (&iter);
+
+      // ... ensure it matches one of the supported mechanism property fields.
+      for (const supported_mechanism_properties *prop = supported_properties; prop->name; ++prop) {
+         // Authentication spec: naming of mechanism properties MUST be case-insensitive. For instance, SERVICE_NAME and
+         // service_name refer to the same property.
+         if (strcasecmp (key, prop->name) == 0) {
+            const bson_type_t type = bson_iter_type (&iter);
+
+            if (type == prop->type) {
+               goto found_match; // Matches both key and type.
+            } else {
+               // Connection String spec: Any invalid Values for a given key MUST be ignored and MUST log a WARN level
+               // message.
+               MONGOC_URI_ERROR (error,
```

Comment:
Returning an error (here and below) contradicts the comments:

> Connection String spec: Any invalid Values for a given key MUST be ignored and MUST log a WARN level message.

Suggest referencing the auth spec as rationale:

> Drivers SHOULD raise an error as early as possible when detecting invalid values in a credential.

Related: DRIVERS-2056 proposes revising the connection string spec requirement to warn.